Python/fileutils.c

```diff
@@ -2096,6 +2107,124 @@ _Py_abspath(const wchar_t *path, wchar_t **abspath_p)
 #endif
 }
 
+// The Windows Games API family does not provide these functions
+// so provide our own implementations. Remove them in case they get added
+// to the Games API family
+// Note that this implementation does not handle all the same cases as the real
+// function, but we expect games are very unlikely to encounter the more obscure
+// cases.
+#if defined(MS_WINDOWS_GAMES) && !defined(MS_WINDOWS_DESKTOP)
+HRESULT
+PathCchSkipRoot(const wchar_t *path, const wchar_t **rootEnd)
+{
+    if (path[0] == '\\') {
+        /* relative path with root e.g. \Windows */
+        if (path[1] != '\\') {
+            *rootEnd = path + 1;
+            return S_OK;
+        }
+
+        /* UNC drives e.g. \\server\share or \\?\UNC\server\share */
+        const wchar_t *end = path + 2;
+        if (!wcsnicmp(end, L"?\\UNC\\", 6)) {
+            end += 6;
+        }
+
+        end = wcschr(end, '\\');
+        if (!end) {
+            *rootEnd = path + wcslen(path);
+            return S_OK;
+        }
+        end = wcschr(end + 1, '\\');
+        *rootEnd = (!end) ? path + wcslen(path) : end + 1;
+        return S_OK;
+    }
+    /* absolute / relative path with drive, e.g. C: or C:\ */
+    else if (isalpha(path[0]) && path[1] == ':') {
+        *rootEnd = (path[2] == '\\') ? path + 3 : path + 2;
+        return S_OK;
+    }
+
+    /* relative path */
+    return E_INVALIDARG;
+}
+
+static HRESULT
+PathCchStripToRoot(wchar_t *path, size_t size)
+{
+    wchar_t *end;
+    if (PathCchSkipRoot(path, &end) == S_OK) {
+        if (*end == '\0') {
+            return S_FALSE;
+        }
+        *end = '\0';
+    }
+
+    return E_INVALIDARG;
+}
+
+static wchar_t*
+PathAddBackslashW(wchar_t *path)
+{
+    size_t len;
+    if (!path) {
+        return NULL;
+    }
+    len = wcslen(path);
+    if (len && path[len - 1] != '\\') {
+        path[len++] = '\\';
+        path[len] = '\0';
+    }
+    return path + len;
+}
+
+#ifndef PATHCCH_ALLOW_LONG_PATHS
+#define PATHCCH_ALLOW_LONG_PATHS 0x01
+#endif
+
+static HRESULT
+PathCchCombineEx(wchar_t *buffer, size_t bufsize, const wchar_t *dirname,
+                 const wchar_t *relfile, unsigned long flags)
+{
+    (void)flags;
+
+    if ((isalpha(relfile[0]) && relfile[1] == ':') ||
+       (relfile[0] == '\\' && relfile[1] == '\\'))
+    {
+        dirname = relfile;
+        relfile = NULL;
+    }
+
+    size_t dir_len = wcslen(dirname);
+    size_t file_len = relfile ? wcslen(relfile) : 0;
+    /* path is at max dirname + filename + backslash + \0 */
+    size_t new_len = dir_len + file_len + 2;
+    if (bufsize >= MAXPATHLEN || new_len > bufsize) {
```

Comment:
The value of `bufsize` isn't limited. The result length may be limited. If `PATHCCH_ALLOW_LONG_PATHS` isn't set in `flags`, and `new_len >= MAX_PATH` (not `MAXPATHLEN`), the returned error should be `HRESULT_FROM_WIN32(ERROR_FILENAME_EXCED_RANGE)`. 

Else if `new_len` is greater than `bufsize`, the returned error should be `HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER)`.

`PATHCCH_ALLOW_LONG_PATHS` also determines whether to convert to or from an extended path, depending on whether the current process supports long paths (Python 3.6+ does). But it's okay to omit that behavior for `GAMES`.

Reply:
Oh yeah, we shouldn't have references to MAXPATHLEN in here.

I'm not so concerned about error codes. They're only going to be reported to users, and really we ought to be getting the buffer right ourselves before calling.

I think it's fine for this implementation to assume `PATHCCH_ALLOW_LONG_PATHS`. Don't worry about mimicking every possible behaviour of the original.

Reply:
```suggestion
    if (new_len > bufsize) {
```
---

> Don't worry about mimicking every possible behaviour of the original.

I agree, if getting the correct behavior is excessively complicated with no tangible benefit. What about my other comment on this function? The way `PathCchStripToRoot()` is used, when `relfile` is a rooted path, leads to incorrect and inconsistent behavior. If I know I'm combining a rooted path, I may use a smaller buffer because I know that most of `dirname` won't be copied -- just the drive part. The current implementation requires using a buffer that's big enough to hold `dirname`, `relname` and a joining backslash. OTOH, if `relname` has a drive or UNC share and thus replaces `dirname`, then the length of the passed-in `dirname` is ignored.

Reply:
Firstly, the suggested change is good.

I'm not particularly worried about the case where `relname` would shorten the overall drive. It seems something you _could_ take advantage of, but generally we don't control either half of the paths that will come in here, which means the caller is going to have to assume worst case of plain-old concatenation.

Maybe just add a comment like `/* assuming worst case result length, so our caller should as well */`.

The only reasonable alternative (without complicating everything) would involve modifying the result buffer before potentially returning an error, which I think is worse. I'm pretty sure the original API uses temporary buffers internally, but there's no real need to go there.

Reply:
Yes I would assume the implementation to use something like `PathAllocCombine` internally

Reply:
If we really wanted to I could come up with an implementation that behaves closer to `PathCchCombineEx`. I was simply not sure it was worth the complexity (I kind of hope that at some point microsoft makes these APIs available on the xbox, since there does not appear to be any inherent reason to provide them on every platform except the xbox)

Reply:
> The only reasonable alternative (without complicating everything) would involve modifying the result buffer before potentially returning an error, which I think is worse. I'm pretty sure the original API uses temporary buffers internally, but there's no real need to go there.

It would calculate `dir_len` based on `PathCchSkipRoot()` if `relfile` begins with a backslash. Then it's the same initial test based on `new_len = dir_len + file_len + 2`, requiring `new_len < bufsize`. Copying `dirname` would now use `wcsncpy(buffer, dirname, dir_len)`, and `PathCchStripToRoot()` would not have to be implemented. So you get better behavior AND it's simpler overall.

Reply:
Hold off on any more changes on this function for a day or two - I'm having some interesting conversations with colleagues right now. Might have a more suitable approach (or at least more information about it).